pkg/report/sarif.go

```diff
@@ -0,0 +1,190 @@
+
+	ruleIndexes := map[string]int{}
+	for _, res := range report.Results {
+		for _, vuln := range res.Vulnerabilities {
+			fullDescription := vuln.Description
+			if fullDescription == "" {
+				fullDescription = vuln.Title
+			}
+			path := vuln.PkgPath
+			if path == "" {
+				path = res.Target
+			}
+
+			sw.addSarifResult(&sarifData{
+				vulnerabilityId:  vuln.VulnerabilityID,
+				severity:         vuln.Severity,
+				url:              vuln.PrimaryURL,
+				resourceType:     res.Type,
+				artifactLocation: toPathUri(path),
+				resultIndex:      getRuleIndex(vuln.VulnerabilityID, ruleIndexes),
+				fullDescription:  html.EscapeString(fullDescription),
+				helpText: fmt.Sprintf("Vulnerability %v\nSeverity: %v\nPackage: %v\nFixed Version: %v\nLink: [%v](%v)\n%v",
+					vuln.VulnerabilityID, vuln.Severity, vuln.PkgName, vuln.FixedVersion, vuln.VulnerabilityID, vuln.PrimaryURL, vuln.Description),
+				helpMarkdown: fmt.Sprintf("**Vulnerability %v**\n| Severity | Package | Fixed Version | Link |\n| --- | --- | --- | --- |\n|%v|%v|%v|[%v](%v)|\n\n%v",
+					vuln.VulnerabilityID, vuln.Severity, vuln.PkgName, vuln.FixedVersion, vuln.VulnerabilityID, vuln.PrimaryURL, vuln.Description),
+				message: fmt.Sprintf("Package: %v\nInstalled Version: %v\nVulnerability %v\nSeverity: %v\nFixed Version: %v\nLink: [%v](%v)",
+					vuln.PkgName, vuln.InstalledVersion, vuln.VulnerabilityID, vuln.Severity, vuln.FixedVersion, vuln.VulnerabilityID, vuln.PrimaryURL),
+			})
+		}
+		for _, misconf := range res.Misconfigurations {
+			sw.addSarifResult(&sarifData{
+				vulnerabilityId:  misconf.ID,
+				severity:         misconf.Severity,
+				url:              misconf.PrimaryURL,
+				resourceType:     res.Type,
+				artifactLocation: toPathUri(res.Target),
+				resultIndex:      getRuleIndex(misconf.ID, ruleIndexes),
+				fullDescription:  html.EscapeString(misconf.Description),
+				helpText: fmt.Sprintf("Misconfiguration %v\nType: %s\nSeverity: %v\nCheck: %v\nMessage: %v\nLink: [%v](%v)\n%s",
+					misconf.ID, misconf.Type, misconf.Severity, misconf.Title, misconf.Message, misconf.ID, misconf.PrimaryURL, misconf.Description),
+				helpMarkdown: fmt.Sprintf("**Misconfiguration %v**\n| Type | Severity | Check | Message | Link |\n| --- | --- | --- | --- | --- |\n|%v|%v|%v|%s|[%v](%v)|\n\n%v",
+					misconf.ID, misconf.Type, misconf.Severity, misconf.Title, misconf.Message, misconf.ID, misconf.PrimaryURL, misconf.Description),
+				message: fmt.Sprintf("Artifact: %v\nType: %v\nVulnerability %v\nSeverity: %v\nMessage: %v\nLink: [%v](%v)",
+					res.Target, res.Type, misconf.ID, misconf.Severity, misconf.Message, misconf.ID, misconf.PrimaryURL),
+			})
+		}
+	}
+	sarifReport.AddRun(sw.run)
+	return sarifReport.PrettyWrite(sw.Output)
+}
+
+func toSarifRuleName(vulnerabilityType string) string {
+	switch vulnerabilityType {
+	case vulnerability.Ubuntu, vulnerability.Alpine, vulnerability.RedHat, vulnerability.RedHatOVAL,
```

Comment:
I guess this is not vulnerability type, but target type. This comes from report.Result.Type, right? If it is the case, these consts should be them.
https://github.com/aquasecurity/fanal/blob/102e2bce224028a9736a5a0d53e9ff20a2c00cc1/analyzer/os/const.go#L5-L57

Reply:
yes, you're right. It seems that is target type. changed.

Reply:
also I removed several tests from here, because there is no os with `-oval` suffix (`oracle-oval`, `debian-oval`, `redhat-oval` etc).
https://github.com/aquasecurity/trivy/blob/76249bdcf0905fb25e74033bd9044f54eb4c1256/pkg/report/writer_internal_test.go#L10-L102

Reply:
`suse-cvrf` and `opensuse-cvrf` were removed too